File: user_service/app/utils/validation.py
import re
from typing import List
from app.config import settings
from app.exceptions import PasswordPolicyError, UserValidationError

def validate_password_strength(password: str) -> None:
    """Validate password against security policy"""
    errors = []

    if len(password) < settings.MIN_PASSWORD_LENGTH:
        errors.append(f"Password must be at least {settings.MIN_PASSWORD_LENGTH} characters long")

    if len(password) > settings.MAX_PASSWORD_LENGTH:
        errors.append(f"Password must be no more than {settings.MAX_PASSWORD_LENGTH} characters long")

    if not re.search(r'[a-zA-Z]', password):
        errors.append("Password must contain at least one letter")

    if settings.REQUIRE_NUMBERS and not re.search(r'\d', password):
        errors.append("Password must contain at least one number")

    if errors:
        raise PasswordPolicyError("; ".join(errors))

def validate_email_domain(email: str) -> None:
    """Validate email domain (basic check)"""
    # Extract domain from email
    domain = email.split('@')[1].lower()
    
    # List of common disposable email domains (in production, use a more comprehensive list)
    disposable_domains = [
        '10minutemail.com', 'tempmail.org', 'guerrillamail.com',
        'mailinator.com', 'throwaway.email', 'temp-mail.org'
    ]
    
    if domain in disposable_domains:
        raise UserValidationError("Disposable email addresses are not allowed")

def sanitize_input(text: str) -> str:
    """Sanitize user input to prevent XSS and other attacks"""
    if not text:
        return text
    
    # Remove potentially dangerous characters
    dangerous_chars = ['<', '>', '"', "'", '&', ';', '(', ')', 'script', 'javascript', 'vbscript']
    
    for char in dangerous_chars:
        text = text.replace(char, '')
    
    return text.strip()

File: frontend/src/utils/validation.ts
// Validation rules matching backend validation
export interface ValidationResult {
  isValid: boolean;
  errors: string[];
}

// Password validation settings (matching backend config)
const MIN_PASSWORD_LENGTH = 8;
const MAX_PASSWORD_LENGTH = 128;

export const validatePasswordStrength = (
  password: string,
): ValidationResult => {
  const errors: string[] = [];

  if (password.length < MIN_PASSWORD_LENGTH) {
    errors.push(
      `Password must be at least ${MIN_PASSWORD_LENGTH} characters long`,
    );
  }

  if (password.length > MAX_PASSWORD_LENGTH) {
    errors.push(
      `Password must be no more than ${MAX_PASSWORD_LENGTH} characters long`,
    );
  }

  if (!/[a-zA-Z]/.test(password)) {
    errors.push("Password must contain at least one letter");
  }

  if (!/\d/.test(password)) {
    errors.push("Password must contain at least one number");
  }

  return {
    isValid: errors.length === 0,
    errors,
  };
};

export const validateEmailDomain = (email: string): ValidationResult => {
  const errors: string[] = [];

  if (!email.includes("@")) {
    errors.push("Invalid email format");
    return { isValid: false, errors };
  }

  const domain = email.split("@")[1]?.toLowerCase();

  if (!domain) {
    errors.push("Invalid email format");
    return { isValid: false, errors };
  }

  // List of common disposable email domains
  const disposableDomains = [
    "10minutemail.com",
    "tempmail.org",
    "guerrillamail.com",
    "mailinator.com",
    "throwaway.email",
    "temp-mail.org",
  ];

  if (disposableDomains.includes(domain)) {
    errors.push("Disposable email addresses are not allowed");
  }

  return {
    isValid: errors.length === 0,
    errors,
  };
};

export const validateEmail = (email: string): ValidationResult => {
  const errors: string[] = [];

  // Basic email format validation
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  if (!emailRegex.test(email)) {
    errors.push("Please enter a valid email address");
  }

  return {
    isValid: errors.length === 0,
    errors,
  };
};
